562 - IdsDatePicker part of remaining features/issues

andriivarhanov opened the pull request:
**Explain the details for making this change. What existing problem does the pull request solve?**
This PR includes part of Date picker and Time picker TODO items from https://github.com/infor-design/enterprise-wc/issues/562 and https://github.com/infor-design/enterprise-wc/issues/636.

**Related github/jira issue (required)**:
Part of https://github.com/infor-design/enterprise-wc/issues/562
Part of https://github.com/infor-design/enterprise-wc/issues/636

**Steps necessary to review your pull request (required)**:
Each TODO item and its description:
> Add functional tests for these cases https://main-enterprise.demo.design.infor.com/components/datepicker/example-anniversary-format.html and https://main-enterprise.demo.design.infor.com/components/datepicker/example-custom-format.html

Not only tests have been added but also adjustments for locale `parseDate` `formatDate` functions to be able to parse/format dates with wide/abbreviated months. Tests added for `ids-locale` as well as for `ids-date-picker`.
- go to http://localhost:4300/ids-date-picker/sandbox.html
- check first column of examples
- see date picker supports various custom formats
***
> Range picker https://main-enterprise.demo.design.infor.com/components/datepicker/example-range.html including https://main-enterprise.demo.design.infor.com/components/datepicker/test-week-picker.html https://github.com/infor-design/enterprise/blob/main/src/components/datepicker/datepicker.js#L69-L79 IdsDatePicker: Additional Features #538

Added ability to select a week to existing range picker.
- go to http://localhost:4300/ids-date-picker/range.html
- check "Range - Week Picker" example
***
> Add disabled dates https://main-enterprise.demo.design.infor.com/components/datepicker/example-disabled-dates.html for month view as well. Can rethink the API see current settings https://github.com/infor-design/enterprise/blob/main/src/components/datepicker/datepicker.js#L40-L57

Couldn't rethink the API, used the ones from enterprise version.
- go to http://localhost:4300/ids-date-picker/disabled-dates.html
- check all examples on the page
- go to http://localhost:4300/ids-date-picker/range.html
- check "Range with disabled dates and legend" example
- go to http://localhost:4300/ids-date-picker/sandbox.html
- check "Available Date Validation" example
***
> Add sizes and field height examples like input https://main-enterprise.demo.design.infor.com/components/datepicker/example-sizes.html with tests

- go to http://localhost:4300/ids-date-picker/sizes.html
- check all width/height sizes added
***
> Date picker with time https://main-enterprise.demo.design.infor.com/components/datepicker/example-timeformat.html with tests see settings https://github.com/infor-design/enterprise/blob/main/src/components/datepicker/datepicker.js#L26-L31

Completed below item from https://github.com/infor-design/enterprise-wc/issues/636
> Separate out the "picker contents" to a separate, embeddable form element (could potentially be shared by this component and the Date Picker's "Date/Time" view)

what allows to embed time picker dropdowns without input.

Completed below item from https://github.com/infor-design/enterprise-wc/issues/636
> Make the picker popup's dropdowns display values that reflect what's in the IdsTriggerField

what allows to parse input string and populate time picker dropdowns with values.

Added some missed items to time picker TODOs.

- go to http://localhost:4300/ids-date-picker/sandbox.html
- check "Date Time Picker AM/PM" and "Date Time Picker 24-Hour" example
- change `minute-interval` and `second-interval` attributes or `minuteInterval` and `secondInterval` settings and time picker dropdowns
***
> Add tests for invalid dates https://main-enterprise.demo.design.infor.com/components/datepicker/example-validation.html

Added tests for valid date/available date validation and examples.
- go to http://localhost:4300/ids-date-picker/sandbox.html
- check "Valid Date Validation" and "Available Date Validation" examples
***
> Add test that the change and input event is fired when the date is changed as a test. And make sure the hide/show events are fired as well (from the popup component). And add to breaking changes.

Added tests, added description to the README
***
> Add test that the placeholder works (should work)

Changed placeholder functionality from custom text to `boolean` attribute, if true, it's showing `format` as the input placeholder

- go to http://localhost:4300/ids-date-picker/example.html
- check "Date Field (required)" example
- go to http://localhost:4300/ids-date-picker/sandbox.html
- check all of the examples have placeholder
***
> Add test for UTC picker https://github.com/infor-design/enterprise/blob/main/src/components/datepicker/datepicker.js#L83

- go to http://localhost:4300/ids-date-picker/sandbox.html
- check "UTC Date" example
- e2e test added as well

I'm not quite sure if I did it right, please leave your feedback
***
> Speed -> if you run the performance tab https://main-enterprise.demo.design.infor.com/components/datepicker/example-index.html is about 10 times slower rendering the page

- pull and run enterprise version https://github.com/infor-design/enterprise
- go to http://localhost:4000/components/datepicker/example-index.html
- run performance tab
- pull and run wc version
- go to http://localhost:4300/ids-date-picker/example.html
- run performance tab
- see wc version values lower than enterprise version values
***

**Included in this Pull Request**:
- [x] An e2e or functional test for the bug or feature.
~- [ ] A note to the change log.~

tmcconechy commented on src/components/ids-month-view/ids-month-view-types.ts:
```diff
@@ -0,0 +1,32 @@
+import type IdsMonthView from './ids-month-view';
+
+export type IdsRangeSettings = {
```
We had decided to put these types in the top of the js file for it - so these would go ids-month-view.js (not a separate file) - just as a common guideline

andriivarhanov replied:
makes sense, moved these types to the js file

tmcconechy reviewed (changes requested):
Looks good - just some minor things. And in terms of functionality i only see one thing (question)

- on http://localhost:4300/ids-date-picker/disabled-dates.html. 
- disable years -> should it disable in the year/month dropdown?

- also on the year/month dropdown would like to. see the new designs https://www.figma.com/file/y1oT4NPGJA78EzUNssZYsG/Calendar?node-id=1%3A2269&viewport=-288%2C2528%2C0.32999536395072937


UTC Date seems ok -> we never really had an example. Could add one without time (just the date as that seems more common)

❤️  The performance fixes

andriivarhanov commented:
@tmcconechy thanks for your feedback

> a) on http://localhost:4300/ids-date-picker/disabled-dates.html then disable years -> should it disable in the year/month dropdown?

It looks like there is no such functionality in the enterprise version, could be a new feature for this version. I've added disabled years to picklist as a test, http://localhost:4300/ids-date-picker/disabled-dates.html "Disable years" example

> b) also on the year/month dropdown would like to see the new designs

I just making the PR with only a part of TODO items since the issue is big and I decided to split. If possible I would push the new designs in the next PR

> c) UTC Date seems ok -> we never really had an example. Could add one without time (just the date as that seems more common)

Replaced UTC Date example with ISO Date example, it's a date without time

tmcconechy commented:
yeah no problem @andriivarhanov this is big. what I can do is make a new issue with whats left after this is merged. Did you want to just mark off all the ones that are done for me and i'll move it to new issues(s)

Thanks!
